File: python/树/1123. 最深叶节点的最近公共祖先.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def lcaDeepestLeaves(self, root: TreeNode) -> TreeNode:
        def dfs(root):
            if not root: return 0
            return 1+max(dfs(root.left),dfs(root.right))
        left = dfs(root.left)
        right = dfs(root.right)
        if left == right:
            return root
        elif left > right:
            return root.left
        else:
            return root.right

File: go/tree/1123. 最深叶节点的最近公共祖先.go
package tree

func lcaDeepestLeaves(root *TreeNode) *TreeNode {
	var dfs func(*TreeNode) int
	dfs = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		left := dfs(root.Left)
		right := dfs(root.Right)
		if left > right {
			return 1 + left
		} else {
			return 1 + right
		}
	}

	left, right := dfs(root.Left), dfs(root.Right)
	if left == right {
		return root
	} else if left > right {
		return lcaDeepestLeaves(root.Left)
	} else {
		return lcaDeepestLeaves(root.Right)
	}
}
